File: Backend/src/main/java/com/CrowdHaven/Backend/services/RoleService.java
package com.CrowdHaven.Backend.services;

import com.CrowdHaven.Backend.DTOS.RoleDTO;
import com.CrowdHaven.Backend.models.Community;
import com.CrowdHaven.Backend.models.Role;
import com.CrowdHaven.Backend.repositories.CommunityRepository;
import com.CrowdHaven.Backend.repositories.RoleRepository;
import com.CrowdHaven.Backend.repositories.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class RoleService {

    private final RoleRepository roleRepository;
    private final UserRepository userRepository;
    private final CommunityRepository communityRepository;

    // Obtener todos los roles de una comunidad
    public List<Role> getRolesByCommunityId(Long communityId) {
        Optional<Community> communityOptional = this.communityRepository.findById(communityId);
        if (communityOptional.isEmpty()) {
            throw new IllegalArgumentException("Comunidad no encontrada");
        }


        return roleRepository.findByCommunityId(communityId);
    }



    public Role addRoleToCommunity(RoleDTO roleDTO) {
        Optional<Community> communityOptional = this.communityRepository.findByName(roleDTO.getCommunity());
        if (communityOptional.isEmpty()) {
            throw new IllegalArgumentException("Comunidad no encontrada");
        }

        Role role = new Role();
        role.setRoleName(roleDTO.getRoleName());
        role.setCommunity(communityOptional.get());

        return roleRepository.save(role);
    }

    public Role updateRole(Long id, RoleDTO roleDTO) {
        Role existingRole = roleRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Rol no encontrado"));

        existingRole.setRoleName(roleDTO.getRoleName());

        return roleRepository.save(existingRole);
    }



    public void deleteRole(Long id) {
        roleRepository.deleteById(id);
    }
}

File: CrowdHaven-front/src/app/api/services/role/role.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Role } from '../../models/role.model';
import { RoleDTO } from '../../dtos/role-dto';

@Injectable({
  providedIn: 'root',
})
export class RoleService {
  private apiUrl = 'http://localhost:8080/CrowdHaven/roles';

  constructor(private http: HttpClient) {}

  getRolesByCommunity(communityId: number): Observable<Role[]> {
    return this.http.get<Role[]>(`${this.apiUrl}/community/${communityId}`);
  }

  addRoleToCommunity(roleDTO: RoleDTO): Observable<RoleDTO> {
    return this.http.post<RoleDTO>(this.apiUrl, roleDTO);
  }

  updateRole(id: number, roleDTO: RoleDTO): Observable<Role> {
    return this.http.put<Role>(`${this.apiUrl}/${id}`, roleDTO);
  }

  deleteRole(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
